runtime/lua/vim/ui/img.lua

```diff
@@ -0,0 +1,486 @@
+---end)
+---```
+---
+---@param opts? {base64?:boolean, size?:integer}
+---@return Iter
+function M:chunks(opts)
+  opts = opts or {}
+
+  -- Chunk size, defaulting to 4k
+  local chunk_size = opts.size or 4096
+
+  local data = self.data
+  if not data or data == '' then
+    return vim.iter(function()
+      return nil, nil, nil
+    end)
+  end
+
+  if opts.base64 then
+    data = vim.base64.encode(data)
+  end
+
+  local pos = 1
+  local len = string.len(data)
+
+  return vim.iter(function()
+    -- If we are past the last chunk, this iterator should terminate
+    if pos > len then
+      return nil, nil, nil
+    end
+
+    -- Get our next chunk from [pos, pos + chunk_size)
+    local end_pos = pos + chunk_size - 1
+    local chunk = data:sub(pos, end_pos)
+
+    -- If we have a chunk available, mark as such
+    local last = true
+    if string.len(chunk) > 0 then
+      last = not (end_pos + 1 <= len)
+    end
+
+    -- Mark where our current chunk is positioned
+    local chunk_pos = pos
+
+    -- Update our global position
+    pos = end_pos + 1
+
+    return chunk, chunk_pos, last
+  end)
+end
+
+---@private
+---Loads the provider used to display and manage the image.
+---@param opts? {write?:fun(...:string)}
+---@return vim.ui.img.Provider|nil
+function M:__load_provider(opts)
+  return require('vim.ui.img.providers').load(self.__provider, opts)
+end
+
+---Displays the image.
+---```lua
+----- Some predefined image
+---local img = vim.ui.img.new({ ... })
+---
+-----Can be invoked synchronously
+---assert(img:show({ ... }):wait())
+---
+-----Can also be invoked asynchronously
+---img:show({ ... }):on_done(function(err)
+---  -- Do something
+---end)
+---```
+---@param opts? vim.ui.img.Opts
+---@return vim.ui.img._Promise<true>
+function M:show(opts)
+  -- Update the opts to reflect what we should be showing next
+  -- such that future updates can work properly in batch
+  self.__opts = require('vim.ui.img.opts').new(opts)
+
+  return self:__schedule(self.__show, self, self.__opts:into_internal_opts())
+end
+
+---@private
+---@param opts vim.ui.img.InternalOpts
+---@return vim.ui.img._Promise<true>
+function M:__show(opts)
+  local promise = require('vim.ui.img._promise').new({
+    context = 'image.show',
+  })
+
+  local provider = self:__load_provider()
+  if not provider then
+    promise:fail('unable to retrieve provider')
+  else
+    provider
+      .show(self, opts)
+      :on_ok(function(id)
+        self.__id = id
+        promise:ok(true)
+      end)
+      :on_fail(function(show_err)
+        promise:fail(show_err)
+      end)
+  end
+
+  return promise
+end
+
+---Hides the image.
+---```lua
+----- Some predefined image
+---local img = vim.ui.img.new({ ... })
+---
+-----Can be invoked synchronously
+---assert(img:hide():wait())
+---
+-----Can also be invoked asynchronously
+---img:hide():on_done(function(err)
+---  -- Do something
+---end)
+---```
+---@return vim.ui.img._Promise<true>
+function M:hide()
```

Comment:
I think this function should internally check if the image is visible or not, so that hide can be called multiple times, it makes the API more pleasant to use.